bool A
bool B
bool Y
bool X
